### PR TITLE
Adds support for reloader in Promscale.

### PR DESCRIPTION
**Pull request by Harkishen-Singh**
Signed-off-by: Harkishen-Singh <harkishensingh@hotmail.com>

Fixes: https://github.com/timescale/o11y-team-applications/issues/182

## Description

*Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.*

## Merge requirements

Please take into account the following non-code changes that you may need to make with your PR:

- [x] CHANGELOG entry for user-facing changes
- [x] Updated the relevant documentation

### CODE REVIEW

**Review by niksajakovljevic — CHANGES_REQUESTED** [review]
Imho this is over-engineered. The code could be simplified a lot. We only need to reload rules.

**Review by niksajakovljevic — COMMENTED** [review]
Looks good. I'd just like to see some tests (either unit or e2e)

File: pkg/runner/runner.go [hunk 1]
```diff
@@ -137,9 +137,52 @@ func Run(cfg *Config) error {
...
+		rulesCtx, stopRuler := context.WithCancel(context.Background())
+		defer stopRuler()
+
+		manager, rulesReloader, err := rules.NewManager(rulesCtx, prometheus.DefaultRegisterer, client, &cfg.RulesCfg)
+		if err != nil {
+			return fmt.Errorf("error creating rules manager: %w", err)
+		}
+		cfg.APICfg.Rules = manager
+
+		reloadRules = rulesReloader
```
Comment (antekresic) [inline pkg/runner/runner.go:165]: I'd rather you declare the manager var before NewManager than doing this, so you can use `=` vs `:=`.

File: pkg/tests/end_to_end_tests/rules_test.go [hunk 2]
```diff
@@ -89,7 +95,24 @@ func TestRecordingRulesEval(t *testing.T) {
 			return nil
 		})
 		require.NoError(t, manager.ApplyConfig(rulesCfg.PrometheusConfig))
-		require.NoError(t, manager.Run(), "error running rules manager") // This is blocking. It will be released after stopRuler() in defer func.
+
+		go func() {
+			require.NoError(t, manager.Run(), "error running rules manager")
```
Comment (antekresic) [inline pkg/tests/end_to_end_tests/rules_test.go:100]: This won't wait for the manager.Run to return a result.

File: pkg/tests/end_to_end_tests/rules_test.go [hunk 3]
```diff
@@ -89,7 +95,24 @@ func TestRecordingRulesEval(t *testing.T) {
...
+		}()
+		wg.Wait() // Wait for first test to complete.
+
+		// Check for rules.
```
Comment (antekresic) [inline pkg/tests/end_to_end_tests/rules_test.go:104]: Try reloading rules before changing them to see if anything changes.

File: pkg/tests/end_to_end_tests/rules_test.go [hunk 4]
```diff
@@ -89,7 +95,24 @@ func TestRecordingRulesEval(t *testing.T) {
...
+		ruleGroups := manager.RuleGroups()
+		require.Equal(t, 1, len(ruleGroups))
+
+		// Reload with an empty configuration file so that there exists no rules.
+		rulesCfg.PrometheusConfigAddress = EmptyRecordingRulesConfigPath
```
Comment (antekresic) [inline pkg/tests/end_to_end_tests/rules_test.go:109]: What if we used empty rules first and then reload to actual rules and wait?

**Review by antekresic — APPROVED** [review]

**Pull request merged**